Add extra message to reserved token receiver percent slider

johnnyd-eth opened the pull request:
## What does this PR do and why?

**Frontend changes:**
1. Clarifies that the reserved token receiver slider percentage is of the overall reserved token allocation
2. Shows the percentage of all newly minted tokens the individual will receive
3. Fixes bug shown below on 'Create Project' by adding text 'project owner' when wallet isn't connect (owner doesn't exist)

**Big code changes:** Moves reserved token receiver modal into a separate component

## Screenshots or screen recordings

https://user-images.githubusercontent.com/96150256/150661332-bb9399b7-1e1f-4db3-94eb-863deb791372.mp4

3. Before:
<img width="613" alt="Screen Shot 2022-01-23 at 11 47 36 am" src="https://user-images.githubusercontent.com/96150256/150661427-7bcded8d-0a01-4c38-be86-2f146a9a157b.png">

After:

<img width="619" alt="Screen Shot 2022-01-23 at 11 45 43 am" src="https://user-images.githubusercontent.com/96150256/150661429-1745d2d5-28f5-4d64-a601-9c8c258ec002.png">

## Acceptance checklist

- [x] I have evaluated the [Approval Guidelines](../CONTRIBUTING.md#approval-guidelines) for this PR.
- [x] I have tested this PR in [all supported browsers](../CONTRIBUTING.md#browser-support).

aeolianeth commented on src/components/shared/inputs/NumberSlider.tsx:
```diff
@@ -43,41 +43,44 @@ export default function NumberSlider({
+          />
+        </Form.Item>
+      </div>
+      <div className="ant-form-item-extra">{formItemProps?.extra ?? ''}</div>
```
**question**: is this preferred to `extra` prop on `Form.Item` in this case?

johnnyd-eth replied:
Yeah I thought so. Because we want the message to apply to the number slider and the input on the right, so span the whole width of the form.

aeolianeth commented on src/components/shared/formItems/ProjectTicketMods.tsx:
```diff
@@ -238,7 +240,7 @@ export default function ProjectTicketMods({
           validator: () => {
             if (total > 100)
-              return Promise.reject('Percentages must add up to 100% or less')
+              return Promise.reject(t`Percentages must add up to 100% or less`)
```
```suggestion
              return Promise.reject(t`Sum of percentages cannot exceed 100%.`)
```

aeolianeth commented on src/components/shared/formItems/ProjectTicketMods.tsx:
```diff
@@ -279,7 +281,10 @@ export default function ProjectTicketMods({
-            <Trans>{100 - total}%to</Trans> <FormattedAddress address={owner} />
+            <Trans>
+              {100 - total}% to{' '}
+              {owner ? <FormattedAddress address={owner} /> : 'project owner'}
```
```suggestion
              {owner ? <FormattedAddress address={owner} /> : t`project owner`}
```

aeolianeth replied:
**question**: I'm wondering how it's possible for `owner` to be undefined here...

johnnyd-eth replied:
On create project when a user doesn't have their wallet connected. See screenshots attached

aeolianeth replied:
oh of course lol. Thanks for explanation 🙏

aeolianeth commented on src/components/modals/ReservedTokenReceiverModal.tsx:
```diff
@@ -0,0 +1,104 @@
+    `${
+      form.getFieldValue('percent')
+        ? ' ' +
+          t`(${(((reservedRate ?? 0) * 1.0 * (percent ?? 0)) / 100).toFixed(2)}%
```
**suggestion**: perhaps would be more readable if we put this in a variable or 2.

```js
const sliderExtra = () => {
  const realTokenAllocation = (reservedRate ?? 0) * (percent ?? 0) * 1.0
  const realTokenAllocationPercent = (realTokenAllocation / 100).toFixed(2)
  const realTokenAllocationText = form.getFieldValue('percent')
    ? ` ${t`(${realTokenAllocationPercent}% of all newly minted tokens)}`}`
    : ''

  return t`The percent this address receives of the project's ${reservedRate}% 
reserved token allocation${realTokenAllocationText}.`
}
```

aeolianeth replied:
Also, why do we need to `* 1.0`? Is this to convert to float or something?

johnnyd-eth replied:
Yeah exactly

aeolianeth replied:
`parseFloat` perhaps?

aeolianeth commented on src/components/Create/ConfirmDeployProject.tsx:
```diff
@@ -197,6 +197,7 @@ export default function ConfirmDeployProject() {
             mods={ticketMods}
             projectId={undefined}
             fundingCycle={undefined}
+            reservedRate={editingFC?.reserved.toNumber() / 2}
```
**question**: Why do we need to divide by 2?

aeolianeth commented on src/components/Dashboard/ReservedTokens.tsx:
```diff
@@ -111,6 +111,7 @@ export default function ReservedTokens({
           mods={ticketMods}
           fundingCycle={fundingCycle}
           projectId={projectId}
+          reservedRate={metadata?.reservedRate / 2} // metadata.reservedRate is double?
```
**question**: Why do we need to divide by 2?

aeolianeth commented on src/components/modals/DistributeTokensModal.tsx:
```diff
@@ -100,6 +100,7 @@ export default function DistributeTokensModal({
               mods={currentTicketMods}
               fundingCycle={currentFC}
               projectId={projectId}
+              reservedRate={metadata?.reservedRate ?? 0}
```
**question**: Why do we need to divide by 2 in the above cases, but not for this case?

aeolianeth replied:
@johnnyd-eth did you end up getting to the bottom of this?

aeolianeth replied:
Oh my bad, this diff is out of date. Ignore me!

aeolianeth reviewed (changes requested):
Nice improvements @johnnyd-eth 💪 . I'll test this locally now, but left a few comments for you in the mean time.

aeolianeth commented on src/components/modals/ReservedTokenReceiverModal.tsx:
```diff
@@ -0,0 +1,104 @@
+          name="beneficiary"
+          defaultValue={form.getFieldValue('beneficiary')}
+          formItemProps={{
+            label: t`Beneficiary`,
```
```suggestion
            label: t`Beneficiary address`,
```

aeolianeth commented on src/components/modals/ReservedTokenReceiverModal.tsx:
```diff
@@ -0,0 +1,104 @@
+          onAddressChange={beneficiary => form.setFieldsValue({ beneficiary })}
+        />
+
+        <Form.Item label={t`Percent`} rules={[{ required: true }]}>
```
```suggestion
        <Form.Item label={t`Percent allocation`} rules={[{ required: true }]}>
```

johnnyd-eth commented:
Thanks for the feedback @aeolianeth, I've addressed all your feedback except the ` * 1.0` -> `parseFloat`, which was causing problems I couldn't be bothered to figure out - I don't see the problem with `* 1.0` as a way to convert int to float.

aeolianeth commented on src/components/modals/ReservedTokenReceiverModal.tsx:
```diff
@@ -0,0 +1,110 @@
+    // and shows the percentage of all newly minted tokens only if percent != 0
+    const realTokenAllocation = (reservedRate ?? 0) * (percent ?? 0) * 1.0
+    const realTokenAllocationPercent = (realTokenAllocation / 100).toFixed(2)
+    let extra =
```
**nitpick (non-blocking)** can be `const`

aeolianeth commented on src/components/Create/TicketingForm.tsx:
```diff
@@ -65,6 +71,7 @@ export default function TicketingForm({
             label: t`Allocate reserved tokens (optional)`,
             extra: t`Automatically distribute a portion of your project's reserved tokens to other Juicebox projects or ETH wallets.`,
           }}
+          reservedRate={reservedRate}
```
**question** is there a reason we can't get the form value here instead:

```suggestion
          reservedRate={form.getFieldValue('reserved')}
```

peripheralist replied:
yes, is the state value necessary here? if it is (cc the recent discussion in discord #frontend) we can maybe just avoid relying on the `Form` entirely to simplify things

johnnyd-eth replied:
https://user-images.githubusercontent.com/96150256/150973885-c6a71f25-b00f-4d7d-a572-9ce672ab2dca.mp4

Without the state, the new value is not updated quickly on the UI. I've removed the other reference to the form in that component, so now it uses the state for everything except the initial load.

aeolianeth commented on src/components/modals/ReconfigureFCModal.tsx:
```diff
@@ -505,6 +505,7 @@ export default function ReconfigureFCModal({
               mods={editingTicketMods}
               projectId={undefined}
               fundingCycle={undefined}
+              reservedRate={fcMetadata?.reservedRate ?? 0}
```
**question** should this also be:

```suggestion
              reservedRate={parseFloat(fromPerbicent(metadata?.reservedRate))}
```

johnnyd-eth replied:
Yep it should be, thanks

aeolianeth reviewed (approved):
Thanks for making those changes @johnnyd-eth 🙇 . Left a couple more non-blockers, so approving and sending to @peripheralist ✅

peripheralist commented:
this is awesome @johnnyd-eth! definitely improves the UX of setting those values. all looks good, though just curious about using that `reservedRate` state alongside the `Form` state and if we can remove one of them. (also very curious about this parseFloat issue lol)

johnnyd-eth commented:
> this is awesome @johnnyd-eth! definitely improves the UX of setting those values. all looks good, though just curious about using that `reservedRate` state alongside the `Form` state and if we can remove one of them. (also very curious about this parseFloat issue)

Thanks for the review @peripheralist! I addressed the conflicting state and form stuff in `TicketingForm`, and added some comments to explain. Also realized neither the `* 1.0` nor parseFloat were necessary for what I was trying to do, so I've removed it.

peripheralist commented:
hey @johnnyd-eth , sorry thought i responded here a few days ago. this looks great! let's just solve the merge conflicts when you get the chance

johnnyd-eth commented:
> hey @johnnyd-eth , sorry thought i responded here a few days ago. this looks great! let's just solve the merge conflicts when you get the chance

Sorted thanks @peripheralist !
